docs/src/main/asciidoc/cdi-integration.adoc

```diff
@@ -0,0 +1,548 @@
+This optimization allows for _framework-level dead code elimination_.
+In few special cases, it's not possible to correctly identify an unused bean. 
+In particular, Quarkus is not able to detect the usage of the `CDI.current()` static method yet.
+Extensions can eliminate possible false positives by producing an `UnremovableBeanBuildItem`.
+
+.`UnremovableBeanBuildItem` Example
+[source,java]
+----
+@BuildStep
+UnremovableBeanBuildItem unremovableBeans() {
+   return UnremovableBeanBuildItem.targetWithAnnotation(STARTUP_NAME); <1>
+}
+----
+<1> Make all classes annotated with `@Startup` unremovable.
+
+== Use Case - My Annotation Is Not Recognized as a Qualifier or an Interceptor Binding
+
+It is likely that the annotation class is not part of the application index.
+For example, classes from the _runtime module_ of a Quarkus extension are not indexed automatically.
+
+_Solution_: Use the `AdditionalBeanBuildItem` as described in <<additional_bean_build_item>>. 
+
+[[annotations_transformer_build_item]]
+== Use Case - I Need To Transform Metadata
+
+In some cases, it's useful to be able to modify the metadata.
+Quarkus provides a powerful alternative to https://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#process_annotated_type[`javax.enterprise.inject.spi.ProcessAnnotatedType`, window="_blank"].
+With an `AnnotationsTransformerBuildItem` it's possible to override the annotations that exist on bean classes.
+
+For example, you might want to add an interceptor binding to a specific bean class.
+Here is how to do it: 
+
+.`AnnotationsTransformerBuildItem` Example
+[source,java]
+----
+@BuildStep
+AnnotationsTransformerBuildItem transform() {
+   return new AnnotationsTransformerBuildItem(new AnnotationsTransformer() {
+
+      public boolean appliesTo(org.jboss.jandex.AnnotationTarget.Kind kind) {
+         return kind == org.jboss.jandex.AnnotationTarget.Kind.CLASS; <1>
+      }
+
+      public void transform(TransformationContext context) {
+         if (context.getTarget().asClass().name().toString().equals("org.acme.Bar")) {
+            context.transform().add(MyInterceptorBinding.class).done(); <2>
+         }
+      }
+    });
+}
+----
+<1> The transformer is only applied to classes.
+<2> If the class name equals to `org.acme.Bar` then add `@MyInterceptorBinding`. Don't forget to invoke `Transformation#done()`.
+
+NOTE: Keep in mind that annotation transformers must be produced _before_ the bean discovery starts.
+
+Build steps can query the transformed annotations for a given annotation target via the `TransformedAnnotationsBuildItem`.
+
+.`TransformedAnnotationsBuildItem` Example
+[source,java]
+----
+@BuildStep
+void queryAnnotations(TransformedAnnotationsBuildItem transformedAnnotations, BuildProducer<MyBuildItem> myBuildItem) {
+   ClassInfo myClazz = ...;
+   if (transformedAnnotations.getAnnotations(myClazz).isEmpty()) { <1>
+     myBuildItem.produce(new MyBuildItem(););
+   }
+}
+----
```

Comment:
I think it would be nice to add a note saying explicitly that the index is not modified.

Reply:
Hmm, this might be a good addition to the `Metadata Sources` paragraph. Generally stating that extensions alter metamodel but not the index itself.

Reply:
Metadata Sources is now a separate section and the jandex index is mentioned there. I believe it's clear enough now ;-).